Feature/bot/bootstrap

convulsiva opened the pull request:
# 🚀 Что сделано
- Добавлена базовая структура проекта:
  - `main.py` — запуск бота  
  - `handlers.py` — обработчики команд  
  - `config.py` — загрузка переменных окружения  
  - `keyboards.py` — клавиатуры  
  - `user_storage.py` — хранение пользователей в JSON  
  - `logging_conf.py` — конфигурация логгера  
- Реализованы команды:
  - `/start` — приветствие и добавление пользователя в базу  
  - `/ping` — проверка доступности бота  
  - `/setgroup` — сохранение группы пользователя  
  - `/today`, `/tomorrow`, `/schedule` — заглушки расписания  
  - `/help` — список всех команд  
- Добавлено хранение пользователей в JSON с поддержкой групп  
- Подключена клавиатура для быстрого вызова команд  
- Настроен базовый логгер для отслеживания ошибок  

---

convulsiva committed:
chore: add project structure and base setup

convulsiva committed:
feat: add minimal working bot with start and ping

convulsiva committed:
feat: add user storage with JSON

convulsiva committed:
feat: integrated help, setgroup commands

convulsiva committed:
fix: fix start command text to avoid HTML parse error

convulsiva committed:
feat: add schedule command with group check and stub data

convulsiva committed:
feat: add JSON user storage with group support

convulsiva committed:
feat: add reply keyboard support in start and button handlers

convulsiva committed:
refactor: improve handlers with russian weekdays and unified schedule logic

convulsiva committed:
feat: add JSON user storage with group support

convulsiva committed:
docs: add TODO

convulsiva committed:
fix: wrap bot start in try/except

convulsiva committed:
feat: validate group number format in /setgroup

convulsiva committed:
refactor: move bot messages to texts.py

convulsiva committed:
refactor: move bot messages to texts.py

convulsiva committed:
fix: wrap bot start in try/except

convulsiva committed:
feat: add easter egg

ErroX1 commented on .env.example:
Перенести нужно этот файл в папку `./bot`, я думаю. Всё что относится к боту в папку `./bot`

convulsiva replied:
сделал a811e23cec9abb5e202c018a575f725f491d2d21

ErroX1 commented on requirements.txt:
Перенести нужно этот файл в папку `./bot`, я думаю. Всё что относится к боту в папку `./bot`

convulsiva replied:
сделал a811e23cec9abb5e202c018a575f725f491d2d21

ErroX1 commented on .env.example:
```diff
@@ -0,0 +1,3 @@
+BOT_TOKEN=
+ADMIN_IDS=
+ENV=dev
```
И добавь пустую строчку в конец файла, пожалуйста 🥺🥺🥺

ErroX1 commented on texts.py:
Перенести нужно этот файл в папку `./bot`, я думаю. Всё что относится к боту в папку `./bot`

convulsiva replied:
сделал f3fec1087cc4cf02ae3a47c9adf7bacf24e1da37

ErroX1 commented on texts.py:
```diff
@@ -0,0 +1,40 @@
+    "/setgroup <номер> - сохранить свою группу\n"
+    "/schedule - показать расписание\n"
+    "/today - расписание на сегодня\n"
+    "/tomorrow - расписание на завтра\n"
```
Нараица. Пока что этого набора команд с головой хватит, в будущем мб добавить команды для привязке пользователя к рассылке, выбор времени для рассылки расписания и тд, там подумаем)

ErroX1 commented on bot/logging_conf.py:
```diff
@@ -0,0 +1,34 @@
+    console_handler = logging.StreamHandler(sys.stdout)
+    console_handler.setFormatter(ColorFormatter(log_format, datefmt=date_format))
+
+    file_handler = logging.FileHandler("bot.log", encoding="utf-8")
```
Может вынести путь до файла с логами в `.env`? 
Например так:
```python 
logging.FileHandler(config("LOG_FILE_NAME"), encoding="utf-8")
```
Посмотрев в .env станет сразу понятно, какие файлы создаются во время работы сервиса и где они будут лежать. 😎

convulsiva replied:
будет сделано🤥

convulsiva replied:
сделал 8bbfb78f763c940fb3d6ea7c138d144b4a2c5319

ErroX1 commented on bot/logging_conf.py:
```diff
@@ -0,0 +1,34 @@
+
+def setup_logging():
+    log_format = "%(asctime)s [%(levelname)s] %(name)s: %(message)s"
+    date_format = "%Y-%m-%d %H:%M:%S"
```
Может тоже в `.env` в одном файле сразу все настройки, удобно будет поменять формат логов.

convulsiva replied:
сделал ac45226509472c5ef69cb95e40ae254fc12fb003

ErroX1 commented on bot/config.py:
Нужно будет выбрать один из 2ух способов загрузки переменных из файла `.env`.
Я загружал переменные окружения через decouple, и мне хватало всего одной строчки:
```python
from decouple import config
BOT_TOKEN = config("BOT_TOKEN")
```

convulsiva replied:
я думаю будем юзать твой, переделаю дам фидбек

ErroX1 commented on bot/services/parser_client.py:
Что будет в этом файле? 🥺

convulsiva replied:
тут я думаю будет клиент для работы с парсером: мы будем создавать объект его класса и через методы получать расписание и прочие данные. типа мост между парсером и ботом

ErroX1 replied:
Глянем, как допишу парсер)

ErroX1 commented on bot/services/user_storage.py:
```diff
@@ -0,0 +1,49 @@
+import json
+import os
+
+FILE_PATH = "storage/users.json"
```
Предлагаю вынести в файл `.env`, ниже об этом будет написано в моих комментах.

convulsiva replied:
сделал d4f5aa9a3c38ee179240e53062cca9cc7f562a30

ErroX1 commented on bot/services/user_storage.py:
```diff
@@ -0,0 +1,49 @@
+
+def load_users():
+    if not os.path.exists(FILE_PATH):
+        return {"users": []}
```
А есть вариант ещё лучше сделать следующую структуру таблицы хранения, а именно хранить по `chad_id` данные о пользователе, ведь `chat_id` у всех разный:
```json
[
{
"12345": {"username": "@applle",
           "first_name": "Pickle",
           "group": "5130999/76454"},
 "54321": {"username": "@orange",
           "first_name": "FallBack",
           "group": "872364/87164"}
}
]
```

convulsiva replied:
сделал 58afdec6831868a78758efcf545c3fd3a7947361

ErroX1 commented on bot/services/user_storage.py:
В целом система хранения данных о пользователях рабочая, единственное, что мне не нравится, так это постоянная перезапись `.json` файла. Я понимаю, что это особенность формата `.json`, ведь его нельзя отредактировать, только перезаписать. Поэтому, в будущем предлагаю добавить новый формат файла (откажемся от `.json`) ИЛИ логику в которой не будет так часто перезаписываться файл. Например, мы будем хранить данные о пользователях в глобальной переменной `USERS_DATA_DICT = dict()`, а сохранять данные в файл `.json` лишь при жёсткой необходимости: перед `fallback`-ом или раз в интервал времени, например, каждые 5 минут.

ErroX1 commented on bot/handlers.py:
```diff
@@ -0,0 +1,120 @@
+        )
+
+    def handle_schedule(message: Message):
+        users = load_users()
```
Тут вовсе читаем файл `.json` при **каждом** сообщении от **одного** пользователя типа: `/schedule` или `🗓 Всё расписание` 😨😨😨

ErroX1 commented on bot/services/user_storage.py:
```diff
@@ -0,0 +1,49 @@
+            "username": username,
+            "first_name": first_name,
+            "group": None
+        })
```
С новой структурой:
```python
users_by_chat_id = load_users()
users_by_chat_id.setdefault(chat_id, {"username": username,
                                      "first_name": first_name,
                                      "group": None})
```

convulsiva replied:
58afdec6831868a78758efcf545c3fd3a7947361

ErroX1 commented on bot/services/user_storage.py:
```diff
@@ -0,0 +1,49 @@
+            "username": None,
+            "first_name": None,
+            "group": group
+        })
```
Аналогично тут было бы:
```python
users_by_chat_id = load_users()
user = users_by_chat_id.setdefault(chat_id, {"username": None,
                                             "first_name": None,
                                             "group": None})
user["group"] = group
```

convulsiva replied:
58afdec6831868a78758efcf545c3fd3a7947361

ErroX1 commented on bot/handlers.py:
```diff
@@ -0,0 +1,120 @@
+
+    @bot.message_handler(func=lambda msg: True)
+    def fallback_handler(message: Message):
+        bot.reply_to(message, UNKNOWN_COMMAND)
```
И добавь пустую строчку в конец файла, пожалуйста 🥺🥺🥺

ErroX1 commented on bot/handlers.py:
```diff
@@ -0,0 +1,120 @@
+from texts import START_MESSAGE, HELP_MESSAGE, INVALID_GROUP_FORMAT, GROUP_SAVED, NOT_REGISTERED, NO_GROUP, PING, DAYS_RU, UNKNOWN_COMMAND, SETGROUP_INSTRUCTION
+import re
+
+def register_handlers(bot: TeleBot):
```
Круто, нравится как загружаются через register_handlers все хендлеры) Удобно 🔥🔥🔥

ErroX1 commented on bot/handlers.py:
```diff
@@ -0,0 +1,120 @@
+    def handle_schedule(message: Message):
+        users = load_users()
+        chat_id = message.chat.id
+        user = next((u for u in users["users"] if u["chat_id"] == chat_id), None)
```
Тут было бы `user = users_by_chat_id.get(chat_id, None)`

convulsiva replied:
58afdec6831868a78758efcf545c3fd3a7947361

ErroX1 commented on bot/handlers.py:
```diff
@@ -0,0 +1,120 @@
+        update_user_group(message.chat.id, group)
+        bot.reply_to(message, GROUP_SAVED.format(group=group))
+
+    @bot.message_handler(func=lambda msg: True)
+    def easter_egg_handler(message: Message):
+        text = message.text.lower()
+
+        triggers = ["serega pirat", "серега пират", "пират"]
+
+        if any(trigger in text for trigger in triggers):
+            bot.send_sticker(
+                message.chat.id,
+                "CAACAgIAAxkBAAOcaMR5j0knrqDF0s1lONeOhEY2syYAAvsYAAIRSwhLT7Bhl6t0YIo2BA"
+            )
```
))) кэфтэмэ, фанаты)

convulsiva replied:
в целом я думаю для целостности также вынести triggers в texts.py

ErroX1 replied:
да, это звучит логично🔥🔥🔥

ErroX1 reviewed (changes requested):
## Отличная работа! ✨💯🤩

### Что понравилось
 - Архитектура огонь, удобно расширить в будущем.
 - Приятные тексты сообщений в боте, по ним всё понятно, что требуется от юзера.

### Что исправить?
 - Вынести некоторые переменные в `.env` файл, удобно от туда наблюдать за всеми настройками проекта
 - Структуру хранения пользователей, чтобы избежать лишних циклов.
 - Мелкие придирки, строчка в конце файла и тд

### Возможно исправить в будущем
 - Сменить `.json` на что-то другое
 - Добавить команды для админов, регистрацию админов и тд.
 - Добавить аннотации типов, было бы удобно

---

Какое-то такое ревью вышло, где-то задушнил, где-то реально надо подправить, в целом всё шикарно, мне нравится! Молодец @convulsiva !

convulsiva committed:
chore: moved requirements.txt and .env.example to the bot folder

convulsiva committed:
chore: moved texts.py to the bot folder

convulsiva committed:
refactor: moved the log path to .env and updated logging_conf

convulsiva committed:
refactor: moved the log format to .env

convulsiva committed:
refactor: moved the path to the users.json file to .env

convulsiva committed:
Refactor(user_storage): Store users by chat_id instead of a list

- users.json now stores users as a dictionary {chat_id: {...}}
- Simplified data searching and updating via dict.get()
- Rewritten add_user and update_user_group
- Updated handlers.py to work with the new structure

ErroX1 reviewed (approved):
Рабочий вариант готов, отлично @convulsiva . Идём в некст PR за функционалом бота) надеюсь уже с моим парсером 😁
